subiquity/common/ubuntu_advantage.py

```diff
@@ -0,0 +1,151 @@
+        implementation. Instead, we create a response based on the following
+        rules:
+        * Tokens starting with "x" will be considered expired.
+        * Tokens starting with "i" will be considered invalid.
+        * Tokens starting with "f" will generate an internal error.
+        """
+        await asyncio.sleep(1 / self.scale_factor)
+
+        if token[0] == "x":
+            path = "examples/uaclient-status-expired.json"
+        elif token[0] == "i":
+            raise InvalidUATokenError(token)
+        elif token[0] == "f":
+            raise CheckSubscriptionError(token)
+        else:
+            path = "examples/uaclient-status-valid.json"
+
+        with open(path, encoding="utf-8") as stream:
+            return json.load(stream)
+
+
+class UAClientUAInterfaceStrategy(UAInterfaceStrategy):
+    """ Strategy that relies on UA client script to retrieve the information.
+    """
+    async def query_info(self, token: str) -> dict:
+        """ Return the subscription info associated with the supplied
+        UA token. The information will be queried using UA client.
+        """
+        command = (
+            "ubuntu-advantage",
+            "status",
+            "--format", "json",
+            "--simulate-with-token", token,
+        )
+        try:
+            proc: CompletedProcess = await utils.arun_command(command,
+                                                              check=True)
+            # TODO check if we're not returning a string or a list
+            return json.loads(proc.stdout)
+        except CalledProcessError:
+            log.exception("Failed to execute command %r", command)
+            # TODO Check if the command failed because the token is invalid.
+            # Currently, ubuntu-advantage fails with the following error when
+            # the token is invalid:
+            # * Failed to connect to authentication server
+            # * Check your Internet connection and try again.
+        except json.JSONDecodeError:
+            log.exception("Failed to parse output of command %r", command)
+
+        message = "Unable to retrieve subscription information."
+        raise CheckSubscriptionError(token, message=message)
+
+
+class UAInterface:
+    """ Interface to obtain Ubuntu Advantage subscription information. """
+    def __init__(self, strategy: UAInterfaceStrategy):
+        self.strategy = strategy
+
+    async def get_subscription(self, token: str) -> dict:
+        """ Return a dictionary containing the subscription information. """
+        return await self.strategy.query_info(token)
+
+    async def get_avail_services(self, token: str) -> list:
+        """ Return a list of available services for the subscription
+        associated with the token provided.
+        """
+        info = await self.get_subscription(token)
+
+        expiration = dt.fromisoformat(info["expires"])
+        if expiration.timestamp() <= dt.utcnow().timestamp():
```

Comment:
Sometimes units have extremely incorrect system times.  Can we run a test with a valid token and system time that is wrong enough to trigger the expiration?  I'd like to know if the u-a client enforces this already.

Reply:
I tried to perform this test but unfortunately, changing the system date leads to TLS certificate validity exception when running `ua attach`. We might need to disable TLS verification in ubuntu-advantage-client or to generate a token that expired just a few days ago.

Reply:
Give this one some more thought.  I lean toward removing this explicit expiration check.

Reply:
I played around with ua attach some more and it seems that the "attach" operation would succeed even if the token is expired. 

I think we have two options:
* removing everything related to token expiration from Subiquity altogether
* changing the workflow to allow the user to go through even if we think the token is expired. That means changing the screen from:
```
Token expired
The Ubuntu Advantage token that you provided has expired.
Please use a different token.
[ Close ]
```
to something like
```
Token expired
The Ubuntu Advantage token that you provided seems to have expired. 
Do you want to go back or continue anyway?
[ Back ] [ Continue anyway ]
```

Do you have a preference?

Reply:
We're due to have a UX discussion on the matter anyhow, so we can discuss as part of that.
Let's go with status-quo until then.